src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CompoundAssignmentOperator.cs

```diff
@@ -507,17 +507,7 @@ private BoundExpression TransformCompoundAssignmentLHS(BoundExpression originalL
                     break;
 
                 case BoundKind.ArrayAccess:
-                    if (isDynamicAssignment)
```

Comment:
Just removing this seems very heavy. 

We should do this only when element type is an unsealed reference type. 
We definitely do not want to do this with struct arrays or we might degrade fairly common patterns like ` intArray[i]++ `

Reply:
Done.

Reply:
Do we have a test that generates / diffs IL to verify the `intArray[i]++` case maintains the current code generation?

Reply:
Added.

Reply:
Since this fix addresses such a corner case, I think the right expectation is that most or none of the existing tests need to change. 
And there should be a lot of existing tests doing `intArray[i]++` or similar.

Reply:
It looks like it is the case after the last commit.